### PR TITLE
Tax office lists

### PR DESCRIPTION
**Pull request by JulianRoesner**
# Short Description
This adds the possibility to request a list of all the tax offices with their `bufa_nr ` grouped by their county.

I'v also changed the structure xml_parser files a little. I am proposing to have a separate folder for xml parsing with one file with the general functions (e.g. find the text of an element with a specific name) and one file with the specific functions to find certain content in specific elster xmls.

# Changes
- Added all the Eric functionalities to get a list of tax offices for a county and to get the list of county ids.
- Add a script to generate a json with all the tax offices
- Add a route to request the tax offices json
- Add the same route to MockErica

# Feedback
- Does the new XML parsing structure make sense to you?
- Do you see any tests missing?

### CODE REVIEW

**Review by JulianRoesner — COMMENTED** [review]
Some scout work 🏕️ 

We still have the following open todo in Eric:
```python
# TODO: Unify usage of EricWrapper; rethink having eric_wrapper as a parameter
@contextmanager
def get_eric_wrapper():
    """This context manager returns an initialised eric wrapper; it will ensure that the ERiC API is shutdown after
    use. """
    eric = EricWrapper()
    with tempfile.TemporaryDirectory() as tmp_dir:
        eric.initialise(log_path=tmp_dir)
        try:
            yield eric
        finally:
            eric.shutdown()
            with open(os.path.join(tmp_dir, 'eric.log'), encoding='latin-1') as eric_log:
                logger.debug(eric_log.read())
```

As far as I can see we use this function in the pyeric controller and in `_generate_transfer_header`. I've changed the usage of `_generate_transfer_header`  to always have the parameter set. However, I do not think we should move the call to `get_eric_wrapper` further up because we would end up in the RequestController and this should not have to deal with the EricWrapper in my head. 

The other possibility for more unification would be to always use the Eric wrapper. However, this is too much overhead for my taste.

File: erica_app/erica/pyeric/eric.py [hunk 1]
```diff
@@ -281,56 +275,94 @@ def create_th(self,
...
+        fun_get_tax_offices = self.eric.EricMtHoleFinanzamtLandNummern
+        argtypes = [c_void_p, c_void_p]
+        restype = int
+
+        return self._run_buffer_method(
+            fun_get_tax_offices,
+            argtypes,
+            restype)
+
+    def _run_buffer_method(self, function, argtypes, restype, *args):
```
Comment (JulianRoesner) [inline erica_app/erica/pyeric/eric.py:350]: Do you have an idea for a better name?

File: erica_app/erica/elster_xml/xml_parsing/erica_xml_parsing.py [hunk 2]
```diff
@@ -10,40 +10,33 @@ def _get_element_from_xml(xml_string: str, element_xpath: str):
...
+def get_elements_text_from_xml_element(xml_element, searched_child_element):
+    matched_elements = xml_element.findall('.//' + searched_child_element)
+    result = []
+    for el in matched_elements:
+        result.append(el.text)
+    if xml_element.tag == searched_child_element:  # also consider the element itself
+        result.append(xml_element.text)
+    return result
+
+
```
Comment (nfelger) [inline erica_app/erica/elster_xml/xml_parsing/erica_xml_parsing.py:39]: I like this, but there's a lot of similar logic in these two functions (and also get_elements_text_from_xml). Maybe we could have a function that does `get_elements_from_xml_element` and the others just call that?. It's more looping but also clearer code. WDYT?

Something like this (untested):
```suggestion
def get_elements_from_xml_element(xml_element, tag_name):
    matched_elements = xml_element.findall('.//' + tag_name)
    result = []
    for el in matched_elements:
        result.append(el)
    if xml_element.tag == element:  # also consider the element itself
        result.append(xml_element)
    return result
    
    
def get_elements_from_xml(xml_string, element):
    xml_tree = remove_declaration_and_namespace(xml_string)
    return get_elements_from_xml_element(xml_tree, element)


def get_elements_text_from_xml_element(xml_element, searched_child_element):
    return [el.text for el in get_elements_from_xml_element(xml_element, searched_child_element)


def get_elements_text_from_xml(xml_string, element):
    xml_tree = remove_declaration_and_namespace(xml_string)
    return [el.text for el in get_elements_from_xml_element(xml_tree, searched_child_element)
```

**Review by nfelger — CHANGES_REQUESTED** [review]
> Does the new XML parsing structure make sense to you?

It does! Nice one.

> Do you see any tests missing?

Sorry I spent too long reviewing the rest, I didn't managed to go through the tests.

File: erica_app/erica/pyeric/eric.py [hunk 3]
```diff
@@ -331,6 +331,29 @@ def process_verfahren(self, xml_string, verfahren, abruf_code=None, transfer_han
         finally:
             self.close_cert_handle(cert_handle)
 
+    def get_tax_offices(self, county_id):
+        """
+        Get all the tax offices for a specific county
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply erica_app/erica/pyeric/eric.py:321]: Yes, state makes so much more sense 🙈

File: erica_app/erica/pyeric/eric.py [hunk 4]
```diff
@@ -281,56 +275,94 @@ def create_th(self,
...
+        function.restype = restype
 
         buf = self.create_buffer()
-        cert_handle = self.get_cert_handle()
         try:
-            res = fun_decrypt_data(self.eric_instance, cert_handle, EricWrapper.cert_pin.encode(), data.encode(), buf)
+            res = function(self.eric_instance, *args, buf)
             check_result(res)
-            logger.debug(f"fun_decrypt_data res {res}")
+            logger.debug(f"function run_buffer res {res}")
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply erica_app/erica/pyeric/eric.py:358]: Nice idea. Love it

File: erica_app/erica/routes.py [hunk 5]
```diff
@@ -109,6 +110,16 @@ def revoke_unlock_code(unlock_code_revocation: UnlockCodeRevocationData, include
         raise HTTPException(status_code=422, detail=e.generate_error_response(include_elster_responses))
 
 
+@app.get(ERICA_VERSION_URL + '/tax_offices/', status_code=status.HTTP_200_OK)
+def get_tax_offices():
+    """
+    The list of tax offices for all counties is requested and returned.
+    """
+    with open("erica/static/tax_offices.json", "r") as json_file:
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply erica_app/erica/routes.py]: Thanks for the hint. This also makes it way easier to set the media type

File: erica_app/erica/request_processing/requests_controller.py [hunk 6]
```diff
@@ -222,3 +221,73 @@ def generate_json(self, pyeric_response: PyericResponse):
         response = super().generate_json(pyeric_response)
         response['address'] = get_address_from_xml(pyeric_response.server_response)
         return response
+
+
+class GetTaxOfficesRequestController:
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply erica_app/erica/request_processing/requests_controller.py:226]: Yes, I was thinking about that do. I left it here because if you look at the abstraction layers from bottom to top (Eric -> routes) this is the abstraction layer where we normally do that.
That said, I see the following two places:
- Create an additional PyericControllor that can be used as if that would be an actual Eric functionality
- Add this to the script as this is the only place where we use it anyway

File: erica_app/scripts/list_permits.py [hunk 7]
```diff
@@ -1,19 +1,24 @@
 import os
 
+import click
+
 from erica.elster_xml import elster_xml_generator
-from erica.elster_xml.elster_xml_parser import remove_declaration_and_namespace
-from erica.pyeric.pyeric_controller import PermitListingPyericController
+from erica.elster_xml.xml_parsing.erica_xml_parsing import remove_declaration_and_namespace
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply erica_app/scripts/list_permits.py:6]: To be honest, they do not work either. I think that we've never used them again after we split Erica and webapp. I've added the same fix as for the `create_tax...`

File: erica_app/erica/pyeric/eric.py [hunk 8]
```diff
@@ -184,21 +184,15 @@ def close_cert_handle(self, cert_handle):
 
     def get_cert_properties(self):
         fun_get_cert_properties = self.eric.EricMtHoleZertifikatEigenschaften
-        fun_get_cert_properties.argtypes = [c_void_p, c_int, c_char_p, c_void_p]
-        fun_get_cert_properties.restype = c_int
-
-        cert_handle = None
-        buffer = None
+        argtypes = [c_void_p, c_int, c_char_p, c_void_p]
+        restype = c_int
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply erica_app/erica/pyeric/eric.py:188]: Yes, I was thinking about setting them directly, too. In the previous approach we used the function directly afterwards. Because the function is now one parameter I wanted to make it explicit that you have to set the argtypes and restype beforehand. 

That said, as you also like that approach more. I'll change it and just add a docstring to the function to explain how to use it.

File: erica_app/erica/request_processing/requests_controller.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (nfelger) [reply erica_app/erica/request_processing/requests_controller.py:226]: I think a PyericController would be a reasonable place as that is the translation layer between ERiC and the Erica domain, right? But the script would also make sense.

File: erica_app/erica/static/tax_offices.json [hunk 9]
```diff
@@ -1 +1 @@
...
...2118"}, {"name": "Finanzamt Kiel", "bufa_nr": "2120"}, {"name": "Finanzamt Lübeck", "bufa_nr": "2122"}, {"name": "Finanzamt Neumünster", "bufa_nr": "2124"}, {"name": "Finanzamt Ostholstein", "bufa_nr": "2125"}, {"name": "Finanzamt Plön", "bufa_nr": "2126"}, {"name": "Finanzamt Ratzeburg", "bufa_nr": "2127"}, {"name": "Finanzamt Rendsburg", "bufa_nr": "2128"}, {"name": "Finanzamt Eckernförde-Schleswig", "bufa_nr": "2129"}, {"name": "Finanzamt Stormarn", "bufa_nr": "2130"}, {"name": "Finanzamt Pinneberg", "bufa_nr": "2131"}]}, {"state_abbrevation": "th", "name": "Thüringen", "tax_offices": [{"name": "Finanzamt Erfurt", "bufa_nr": "4151"}, {"name": "Finanzamt Ilmenau", "bufa_nr": "4154"}, {"name": "Finanzamt Eisenach", "bufa_nr": "4155"}, {"name": "Finanzamt Gotha", "bufa_nr": "4156"}, {"name": "Finanzamt Mühlhausen", "bufa_nr": "4157"}, {"name": "Finanzamt Sondershausen", "bufa_nr": "4159"}, {"name": "Finanzamt Gera", "bufa_nr": "4161"}, {"name": "Finanzamt Jena", "bufa_nr": "4162"}, {"name": "Finanzamt Pößneck", "bufa_nr": "4165"}, {"name": "Finanzamt Altenburg", "bufa_nr": "4166"}, {"name": "Finanzamt Sonneberg", "bufa_nr": "4170"}, {"name": "Finanzamt Suhl", "bufa_nr": "4171"}]}]}
```
Comment (nfelger) [inline erica_app/erica/static/tax_offices.json:1]: Wait, I'm confused… The json file _did_ have correct unicode escapes? Am I misremembering or did your browser show mangled characters when you demoed it yesterday?

IMO unicode escapes (`\u....`) would be generally preferable to encoding values in UTF-8 because it's maximally compatible that way.

**Review by nfelger — APPROVED** [review]

File: erica_app/erica/static/tax_offices.json (hunk 9 above)
Comment (JulianRoesner) [reply erica_app/erica/static/tax_offices.json:1]: It did show mangled characters. I've changed the script to use UTF-8 encoding when generating the file: 

```python
 with open(_TAX_OFFICES_JSON_FILE_NAME, 'w', encoding="utf-8") as tax_offices_file:
         tax_offices_file.write(json.dumps(tax_office_list))
```

As we then also send the file with `media_type="application/json; charset=utf-8"` this will not show mangled characters. We could also store the file different. However, I think this makes it easy to send the file in the correct formatting, don't you think?

Comment (nfelger) [reply erica_app/erica/static/tax_offices.json:1]: Hm, I'm not so keen on jumping through hoops to create a utf-8 JSON files when an ascii version with unicode escapes is arguably better. It seems to me that the previous file was perfectly fine, and that the mangled characters must have had another source?

Don't take my word, try it out (ba913a9127cbc49b41ef2d3020ad2b0173706f60 is the previous version):
```
$ python3
>>> import requests
>>> requests.get('https://raw.githubusercontent.com/digitalservice4germany/steuerlotse/ba913a9127cbc49b41ef2d3020ad2b0173706f60/erica_app/erica/static/tax_offices.json').json()['tax_offices'][0]
{'county_abbrevation': 'bw', 'name': 'Baden-Württemberg', 'tax_offices': [{'name': 'Finanzamt Offenburg Außenstelle Achern', 'bufa_nr': '2801'}, {'name': 'Finanzamt Villingen-Schwenningen Außenstelle Donaueschingen', ...
```

It loads everything as it should. Same thing if I put it in an online tool like https://jsonformatter.org/#.

So I think we should revert to the previous way of generating the file and figure out where the mangled characters got created. (Btw, I believe we shouldn't have to set a `media_type` on the FileResponse as long as the file ends in `.json`: ["media_type - A string giving the media type. If unset, the filename or path will be used to infer a media type."](https://www.starlette.io/responses/#fileresponse)

Comment (JulianRoesner) [reply erica_app/erica/static/tax_offices.json:1]: > It seems to me that the previous file was perfectly fine, and that the mangled characters must have had another source?

I think the source for the mangled characters is the browser because they are also shown if I request `https://raw.githubusercontent.com/digitalservice4germany/steuerlotse/ba913a9127cbc49b41ef2d3020ad2b0173706f60/erica_app/erica/static/tax_offices.json`directly. Firefox will show it correctly because it parses JSON.
`UTF-8` is the default/only encoding for JSON ([s. here](https://datatracker.ietf.org/doc/html/rfc8259#section-8.1)). Therefore, whenever we parse it, it is correctly decoded (in python, as well as Firefox and with the jsonformatter tool). You can see that also below. Safari and Chrome do not seem to be parsing it as JSON but rather just showing it like a text file -> mangled characters.

Basically, we can use the old file, get rid of all the utf-8 encoding loops and it works as before (Browser will show the mangled characters but webapp won't).

```python
$ python3
>>> r = requests.get('https://raw.githubusercontent.com/digitalservice4germany/steuerlotse/ba913a9127cbc49b41ef2d3020ad2b0173706f60/erica_app/erica/static/tax_offices.json')
>>>r.encoding
'utf-8'
```

> Btw, I believe we shouldn't have to set a media_type on the FileResponse as long as the file ends in .json: "media_type - A string giving the media type. If unset, the filename or path will be used to infer a media type."

Yes, you are correct. I've only had to add that because of the browser.

File: erica_app/erica/routes.py [hunk 10]
```diff
@@ -109,6 +109,14 @@ def revoke_unlock_code(unlock_code_revocation: UnlockCodeRevocationData, include
         raise HTTPException(status_code=422, detail=e.generate_error_response(include_elster_responses))
 
 
+@app.get(ERICA_VERSION_URL + '/tax_offices/', status_code=status.HTTP_200_OK)
+def get_tax_offices():
+    """
+    The list of tax offices for all states is requested and returned.
+    """
+    return FileResponse("erica/static/tax_offices.json")
```
Comment (frederike-ramin) [inline erica_app/erica/routes.py:117]: Okay, I feel like I missed something here: Why don't we request the tax_offices at Elster, but instead only have a file?

Comment (JulianRoesner) [reply erica_app/erica/routes.py:117]: The idea is basically to reduce latency
